### PR TITLE
Added a module to create initial configuration of non-overlapping sph…

### PR DESCRIPTION
**Pull request by torrewk**
Added a module to create initial configuration of non-overlapping sphere, up to (about) 0.6 in volume fraction.

### CODE REVIEW

File: jfsd/utils.py [hunk 1]
```diff
@@ -375,16 +294,238 @@ def initialize_single_neighborlist(
...
+            #Compute and add Brownian velocity
+            key, random_noise = generate_random_array(key, (6*N))
+            brow = thermal.compute_BD_randomforce(N, T, dt,random_noise)   
+            net_vel = add_thermal_noise(net_vel,brow)
+            
+            #Update positions
+            new_positions, nbrs, new_displacements, net_vel = update_pos(positions,displacements,net_vel,nbrs)
+            
+            if(nbrs.did_buffer_overflow):
+                print('its happening')
```
Comment (qubixes) [inline jfsd/utils.py:500]: Debug print?

File: jfsd/utils.py [hunk 2]
```diff
@@ -375,16 +294,238 @@ def initialize_single_neighborlist(
...
+                nbrs = allocate_nlist(positions, neighbor_fn)
+                nl = np.array(nbrs.idx)
```
Comment (qubixes) [inline jfsd/utils.py:502]: I guess this is a neighbor list, but it might be nicer to have a bit more descriptive variable names in this section.  Perhaps "k" is too little descriptive as well? Might be confused with k_b.

File: jfsd/utils.py [hunk 3]
```diff
@@ -375,16 +294,238 @@ def initialize_single_neighborlist(
...
+    phi_eff = N/(L*L*L)*(sigma*sigma)*np.pi/3
+    print("Effective vol fraction is ", phi_eff)
+    if(phi_eff > 0.65):
+                raise ValueError(
+                    "Attempted to create particles configuration too dense. Use imported coordinates instead. Abort!")    
+    print("Overlaps while creating random configurations: ", overlaps, ' Steps to perform are ', Nsteps, ' k is ', k)    
+    while(overlaps>0):
+        start_time = time.time()
+
+        for i in range(Nsteps):
```
Comment (qubixes) [inline jfsd/utils.py:482]: i_step?

File: jfsd/utils.py [hunk 4]
```diff
@@ -375,16 +294,238 @@ def initialize_single_neighborlist(
...
+                displacements = new_displacements
+            
+            
+            
+        k *= 2
+        buffer = (displacements[nl[0, :],nl[1, :],:])
+        buffer = jnp.sqrt(buffer[:,0]*buffer[:,0]+buffer[:,1]*buffer[:,1]+buffer[:,2]*buffer[:,2])
+        print(np.sort(buffer[buffer != 0]))
+        overlaps = check_overlap(displacements,2.)
+        print("Overlaps after some thermalization: ", overlaps, ' TPS are ',Nsteps/(time.time()-start_time),' k is ',k)
```
Comment (qubixes) [inline jfsd/utils.py:522]: TPS -> timesteps per second?

File: jfsd/utils.py [hunk 5]
```diff
@@ -375,16 +294,238 @@ def initialize_single_neighborlist(
...
+    Returns
+    -------
+    nbrs
+    
+    """   
+    #Update neighbor list
+    nbrs = nbrs.update(positions)
+    return nbrs
+
+def CreateHardSphereConfiguration(
```
Comment (qubixes) [inline jfsd/utils.py:345]: I suppose this completely replaces the random configuration? This is a function, so please use snake_case.

File: jfsd/utils.py [hunk 6]
```diff
@@ -375,16 +294,238 @@ def initialize_single_neighborlist(
     neighbor_fn = partition.neighbor_list(displacement,  
                                               jnp.array([Lx, Ly, Lz]),
                                               r_cutoff=space_cut,  # Spatial cutoff for 2 particles to be neighbor
-                                              dr_threshold=0.1,  # displacement of particles threshold to recompute neighbor list
-                                              capacity_multiplier=1,
+                                              #dr_threshold=0.1,  # displacement of particles threshold to recompute neighbor list
+                                              capacity_multiplier=1.1,
```
Comment (qubixes) [inline jfsd/utils.py:298]: Why did it change and why should it be 1.1?

**Review by qubixes — CHANGES_REQUESTED** [review]
I have lightly reviewed the code. I think the code could be improved a little on the readability front. Otherwise, I hope the new code works as it should!

File: jfsd/appliedForces.py [hunk 7]
```diff
@@ -211,8 +211,8 @@ def compute_hs_forces(
 
         #compute forces for each pair
         # spring constant (with lubrication hydrodynamic this should to be ~ o(1000) at least
-        # because of divergent effective drag coeff on particles)
-        k = jnp.where(HIs_flag > 0, 1000. , 100)
+        # because of divergent (at contact) effective drag coeff on particles)
+        k = jnp.where(HIs_flag > 1, 1000. , 100)
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/appliedForces.py:215]: The level of accuracy for hydrodynamic interactions.

File: jfsd/utils.py [hunk 8]
```diff
@@ -375,16 +294,238 @@ def initialize_single_neighborlist(
...
+        Input neighbor lists
+    
+    Returns
+    -------
+    nbrs
+    
+    """   
+    #Update neighbor list
+    nbrs = nbrs.update(positions)
+    return nbrs
```
[the remark this answers is not in the document]
Comment (torrewk) [reply jfsd/utils.py:343]: They are needed to be separate functions (from the main, at least) because these are jit-compiled and run on the GPU. Also, I plan to implement a neighbor list 'machinery' of my own soon, so it's useful to start isolating the part of the code that will need changes (and after these changes, these function will not be that short anymore).

File: jfsd/utils.py (hunk 1 above)
Comment (torrewk) [reply jfsd/utils.py:500]: Forgot to erase. I'm removing this.

File: jfsd/utils.py (hunk 2 above)
Comment (torrewk) [reply jfsd/utils.py:502]: Better documentation/description (and nomenclature) will be implemented soon. Also, in soft matter particle simulations 'k' is often used as a spring constant.

File: jfsd/utils.py (hunk 3 above)
Comment (torrewk) [reply jfsd/utils.py:482]: Modified

File: jfsd/utils.py (hunk 5 above)
Comment (torrewk) [reply jfsd/utils.py:345]: changed the name of the function

File: jfsd/utils.py (hunk 6 above)
Comment (torrewk) [reply jfsd/utils.py:298]: The size of the allocated neighbor list depends on the capacity multiplier. If it's 1, then the list will have the smallest size needed to accommodate all neighbors. This means that once the particle configuration change to a one where the list is not big enough, we need to reallocate it, and this cannot be done on GPU. Thus, to reduce the amount of time we re-allocate the lists, we increase a bit the initial size of the lists.

File: jfsd/appliedForces.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (qubixes) [reply jfsd/appliedForces.py:215]: Okay, I might play with it in the configuration PR.
